File: 7453-합이0인네정수.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())
A = []
B = []
C = []
D = []
AB = []
CD = []

for i in range(n) :
    arr = list(map(int, input().split(' ')))
    A.append(arr[0])
    B.append(arr[1])
    C.append(arr[2])
    D.append(arr[3])

for i in range(n):
    for j in range(n):
        AB.append(A[i] + B[j])
        CD.append(C[i] + D[j])

AB.sort()
CD.sort()

def upper(start, end, arr, target):
    while start < end :
        mid = (start+end) // 2
        if arr[mid] <= target: start = mid + 1
        else: end = mid
    return start
            
def lower(start, end, arr, target):
    while start < end:
        mid = (start+end) // 2
        if arr[mid] < target: start = mid + 1
        else: end = mid
    return start

answer = 0
for i in range(n*n):
    up = upper(0, len(AB)-1, AB, -CD[i])
    low = lower(0, len(AB)-1, AB, -CD[i])
    answer +=  (up - low)

print(answer)

File: js/7453-합이0인네정수.js
const readline = require('readline');

(async () => {
  let rl = readline.createInterface({ input: process.stdin });

  let count = 0;
  let n;
  const A = [],
    B = [],
    C = [],
    D = [];

  for await (const line of rl) {
    if (count === 0) {
      n = +line;
    } else {
      const temp = line.split(' ').map(Number);
      A.push(temp[0]);
      B.push(temp[1]);
      C.push(temp[2]);
      D.push(temp[3]);
    }

    count++;
    if (count >= n + 1) {
      rl.close();
    }
  }

  const arrayAB = [],
    arrayCD = [];

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      arrayAB.push(A[i] + B[j]);
      arrayCD.push(C[i] + D[j]);
    }
  }

  arrayCD.sort((a, b) => a - b);

  const upper = (target, arr) => {
    start = 0;
    end = arr.length;

    while (start < end) {
      let mid = Math.floor((start + end) / 2);

      if (arr[mid] <= target) start = mid + 1;
      else end = mid;
    }
    return start;
  };

  const lower = (target, arr) => {
    start = 0;
    end = arr.length;

    while (start < end) {
      let mid = Math.floor((start + end) / 2);

      if (arr[mid] < target) start = mid + 1;
      else end = mid;
    }
    return start;
  };

  let answer = 0;

  for (let i = 0; i < arrayCD.length; i++) {
    const low = lower(-arrayAB[i], arrayCD);
    const up = upper(-arrayAB[i], arrayCD);
    answer += up - low;
  }

  console.log(answer);
  process.exit();
})();
